openshift-kube-apiserver/filters/deprecatedapirequest/apiaccess_count_controller.go

```diff
@@ -94,24 +95,40 @@ func (c *controller) Start(stop <-chan struct{}) {
 	}()
 
 	// write out logs every c.updatePeriod
-	go wait.NonSlidingUntilWithContext(ctx, c.persistRequestCountForAllResources, c.updatePeriod)
+	go wait.NonSlidingUntilWithContext(ctx, c.sync, c.updatePeriod)
+}
+func (c *controller) sync(ctx context.Context) {
+	currentHour := time.Now().Hour()
+	c.persistRequestCountForAllResources(ctx, currentHour)
 }
 
-func (c *controller) persistRequestCountForAllResources(ctx context.Context) {
-	klog.V(2).Infof("updating top APIRequest counts")
-	defer klog.V(2).Infof("finished updating top APIRequest counts")
+func (c *controller) persistRequestCountForAllResources(ctx context.Context, currentHour int) {
+	klog.V(4).Infof("updating top APIRequest counts")
+	defer klog.V(4).Infof("finished updating top APIRequest counts")
 
 	// get the current count to persist, start a new in-memory count
 	countsToPersist := c.resetRequestCount()
 
 	// remove stale data
-	expiredHour := (time.Now().Hour() + 1) % 24
-	currentHour := time.Now().Hour()
+	expiredHour := (currentHour + 1) % 24
 	countsToPersist.ExpireOldestCounts(expiredHour)
 
 	// when this function returns, add any remaining counts back to the total to be retried for update
 	defer c.requestCounts.Add(countsToPersist)
 
+	// Add resources that have an existing APIRequestCount so that the current and hourly logs
+	// continue to rotate even if the resource has not had a request since the last restart.
+	c.loadOnce.Do(func() {
+		arcs, err := c.client.List(ctx, metav1.ListOptions{})
```

Comment:
There is a side-effect of this that means that it becomes impossible to remove apicounts.  Practically, that's a very unlikely problem, but there needs to be a plan (manual with etcdctl I suspect) to get someone unstuck.

1. 4.9 starts at t=0
2. access ingress.v1beta1
3. write apicount at t0+30m
4. 4.9 stops
5. 4.10 starts t0+2h
6. nobody ever touchest ingress.v1beta1 again
7. t0+6h, cluster-admin deletes apicount/ingress.v1beta1
8. apicount/ingress.v1beta1 gets recreated (before this PR, it did not get recreated)

Reply:
I have a removed the side-effect (step 8 should not happen). See `ApplyStatus()` in `deprecatedapirequest/v1helpers/helpers.go`.